sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py

```diff
@@ -4,150 +4,401 @@
 import numba_dpex as dpex
 
 
-def _make_initialize_window_kernel_funcs(
-    n_clusters,
+def make_pairwise_ops_base_kernel_funcs(
+    n_samples,
     n_features,
-    work_group_size,
-    window_n_centroids,
+    n_clusters,
     window_n_features,
+    window_n_centroids,
+    ops,
     dtype,
+    work_group_size,
     initialize_window_of_centroids_half_l2_norms=False,
 ):
-    zero = dtype(0.0)
-    zero_idx = np.int64(0)
-    inf = dtype(math.inf)
+    # The kernel funcs in this file must behave differently depending on whether the
+    # window over the array of centroids (which has a fixed size):
+    #    - covers a full set of elements of the array (_full window_),
+    #    - covers a partial set of elements (_last window_), which can happen for
+    #      the last windows along each dimension if the size of the window along
+    #      said dimensions does not divide the size of the array of centroids.
+    #
+    # Indeed, for such windows, the loops on the elements of the window must  be
+    # interrupted early to not include compute that would correspond to out-of-bounds
+    # elements in the array of centroids (which would, in this case, cause extra
+    # compute time that, depending on the case, could noticeably hurt performance).
 
-    n_window_features_per_work_group = work_group_size // window_n_centroids
+    # Unfortunately, `numba_dpex` does not offer tools that would enable at the same
+    # time re-using the same kernel functions for the different cases, and passing the
+    # relevant variables (here, `window_n_centroids`, `last_window_n_centroids`,
+    # `window_n_features` and `last_window_n_features`), whose value is known at
+    # compile time, as constant variables (which can benefit performance). This
+    # function proposes a best effort at keeping things the most possibly factorized
+    # and readable, while enabling passing the variables as compile-time constants.
+    # It consists in instanciating a kernel functions for each different value the
+    # variables can take, and adding conditional switches to use the appropriate
+    # instance depending on the state of the main loop over the windows
+    # (`is_last_centroid_window`, `is_last_feature_window`)
 
-    centroids_window_height_ratio_multiplier = (
-        window_n_features // n_window_features_per_work_group
+    last_window_n_centroids = ((n_clusters - 1) % window_n_centroids) + 1
+    last_window_n_features = ((n_features - 1) % window_n_features) + 1
```

Comment:
Those seem correct but I don't understand how you came up with those :)

I would have defined them as:

```suggestion
    last_window_n_centroids = n_clusters % window_n_centroids or window_n_centroids
    last_window_n_features = n_features % window_n_features or window_n_features
```

Reply:
> Those seem correct but I don't understand how you came up with those :)

A twisted mind maybe :x